scripts/canaryRelease.mts

```diff
@@ -0,0 +1,242 @@
+      return [];
+    }
+  }
+
+  if (!baseline) {
+    const { stdout: latestTag } = await $`git describe --abbrev=0`;
+    baseline = latestTag;
+  }
+
+  console.log(`Looking for changed public packages since ${chalk.yellow(baseline)}...`);
+
+  const changedPackages = await listPublicChangedPackages(baseline);
+  if (changedPackages.length === 0) {
+    console.log('Nothing found.');
+  }
+
+  return changedPackages;
+}
+
+async function confirmPublishing(changedPackages: PackageInfo[], yes: boolean) {
+  if (!yes) {
+    const rl = readline.createInterface({
+      input: process.stdin,
+      output: process.stdout,
+    });
+
+    console.log('\nFound changes in the following packages:');
+    for (const pkg of changedPackages) {
+      console.log(` - ${pkg.name}`);
+    }
+
+    console.log('\nThis will publish the above packages to the npm registry.');
+    const answer = await rl.question('Do you want to proceed? (y/n) ');
+
+    rl.close();
+
+    if (answer.toLowerCase() !== 'y') {
+      console.log('Aborted.');
+      process.exit(0);
+    }
+  }
+}
+
+async function setAccessToken(npmAccessToken: string | undefined) {
+  if (!npmAccessToken && !process.env.NPM_TOKEN) {
+    console.error(
+      '❌ NPM access token is required. Either pass it as an argument or set it as an NPM_TOKEN environment variable.',
+    );
+    process.exit(1);
+  }
+
+  const npmrcPath = resolve(workspaceRoot, '.npmrc');
+
+  await appendFile(
+    npmrcPath,
+    `//registry.npmjs.org/:_authToken=${npmAccessToken ?? process.env.NPM_TOKEN}\n`,
+  );
+}
+
+async function setVersion(packages: PackageInfo[]) {
+  const { stdout: currentRevisionSha } = await $`git rev-parse --short HEAD`;
+  const timestamp = formatDate(new Date());
+  let hasError = false;
+
+  const tasks = packages.map(async (pkg) => {
+    const packageJsonPath = resolve(pkg.path, './package.json');
+    try {
+      const packageJson = JSON.parse(await readFile(packageJsonPath, { encoding: 'utf8' }));
+      const version = packageJson.version;
+      const dashIndex = version.indexOf('-');
+      let newVersion = version;
+      if (dashIndex !== -1) {
+        newVersion = version.slice(0, dashIndex);
+      }
+
+      newVersion = `${newVersion}-dev.${timestamp}-${currentRevisionSha}`;
+      packageJson.version = newVersion;
+
+      await writeFile(packageJsonPath, JSON.stringify(packageJson, null, 2) + '\n');
+    } catch (error) {
+      console.error(`${chalk.red(`❌ ${packageJsonPath}`)}`, error);
+      hasError = true;
+    }
+  });
+
+  await Promise.allSettled(tasks);
+  if (hasError) {
+    throw new Error('Failed to update package versions');
+  }
+}
+
+function buildPackages() {
+  if (process.env.CI) {
+    return $$`pnpm build:public:ci`;
+  }
+
+  return $$`pnpm build:public`;
+}
+
+function formatDate(date: Date) {
+  // yyMMddHHmmss
+  return date
+    .toISOString()
+    .replace(/[-:TZ]/g, '')
+    .slice(2, 14);
```

Comment:
Why slice the year? It feels like it breaks the pattern matching for what the string is about.

```suggestion
    .slice(0, 14);
```